File: app/src/lib/modelSettings.js
"use client";

/**
 * The model choices, read from and written to the server.
 *
 * `modelTiers.js` still owns the tier vocabulary and the localStorage copy;
 * this owns the round trip. The two exist together on purpose: the browser
 * copy is what paints the page before the network answers and what a
 * signed-out install runs on, and the server copy is what every run reads —
 * including the scheduled brief, which has no browser at all.
 *
 * When they disagree the server wins. A stale tab that saved a map an hour ago
 * must not overwrite what the user has since chosen on their phone.
 *
 * Every write is a partial: send the one field that changed. A whole-object PUT
 * would let a tab loaded before the switch was flipped write the old value back
 * on the next tier change.
 */

import { BASE } from "./api";
import { authedHeaders, hasAuthToken } from "./authFetch";

const ENDPOINT = `${BASE}/api/user/model-settings`;

/** Fired after a successful save, so other surfaces can re-read. */
export const MODEL_SETTINGS_CHANGED = "duct:model-settings-changed";

export const SETTINGS_DEFAULTS = Object.freeze({
  tiers: {},
  auto_fallback: true,
  engine: "",
});

/**
 * The saved settings, or the defaults.
 *
 * Never throws and never rejects: this gates a settings page, not a run, and a
 * page that fails to paint because a preference endpoint was slow is a worse
 * outcome than a page showing the defaults.
 */
export async function fetchModelSettings() {
  if (!hasAuthToken()) return { ...SETTINGS_DEFAULTS };
  try {
    const res = await fetch(ENDPOINT, { headers: authedHeaders() });
    if (!res.ok) return { ...SETTINGS_DEFAULTS };
    const body = await res.json();
    return {
      tiers: body?.tiers && typeof body.tiers === "object" ? body.tiers : {},
      auto_fallback: body?.auto_fallback !== false,
      engine: String(body?.engine || ""),
    };
  } catch {
    return { ...SETTINGS_DEFAULTS };
  }
}

/**
 * Persist the fields given. Returns what the server now holds, or null when it
 * could not be reached — the caller keeps its optimistic state and says so.
 */
export async function saveModelSettings(patch) {
  if (!hasAuthToken()) return null;
  try {
    const res = await fetch(ENDPOINT, {
      method: "PUT",
      headers: { ...authedHeaders(), "Content-Type": "application/json" },
      body: JSON.stringify(patch || {}),
    });
    if (!res.ok) return null;
    const body = await res.json();
    window.dispatchEvent(new CustomEvent(MODEL_SETTINGS_CHANGED, { detail: body }));
    return body;
  } catch {
    return null;
  }
}

File: backend/routes/model_settings.py
"""The user's model choices: the tier map, and the fallback switch.

Two verbs and one row. The interesting decision is that this exists at all —
the map used to live only in ``localStorage``, which meant the scheduled brief,
the one run whose owner is definitely not watching, could not read the
preference its owner had set. See ``models/settings.py``.

Scoped to the caller by ``get_current_user``: there is no path here that takes
a user id from the request. That is the rule the route-auth boundary test
enforces, and it is the whole security model of this endpoint — the row *is*
the user.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends
from pydantic import BaseModel, ConfigDict

from models.auth import User
from service.auth import get_current_user
from service.model_settings import get_model_settings, save_model_settings

router = APIRouter(tags=["model-settings"])


class ModelSettingsBody(BaseModel):
    """Every field optional — the page saves one control at a time.

    A whole-object PUT would let a tab that loaded before the switch was
    flipped write the old value back on the next tier change.
    """

    model_config = ConfigDict(extra="ignore")

    tiers: dict[str, str] | None = None
    auto_fallback: bool | None = None
    engine: str | None = None


def _payload(settings) -> dict:
    return {
        "tiers": settings.tiers,
        "auto_fallback": settings.auto_fallback,
        "engine": settings.engine,
    }


@router.get("")
def read_model_settings(user: User = Depends(get_current_user)) -> dict:
    return _payload(get_model_settings(user.id))


@router.put("")
def write_model_settings(
    body: ModelSettingsBody, user: User = Depends(get_current_user)
) -> dict:
    return _payload(
        save_model_settings(
            user.id,
            tiers=body.tiers,
            auto_fallback=body.auto_fallback,
            engine=body.engine,
        )
    )
